### DISCUSSION TITLE
새로운 Entity인지 판단하는 방법

### DISCUSSION THREAD

**Discussion by Kimprodp** [2024-woowacourse-study/level-interview#127] [Spring Data]
.

**Comment by jimi567** [comment]
JPA에서 특정 엔티티가 새로운 엔티티인지 판별하기 위해서는 다음과 같은 방법을 사용할 수 있습니다.

1. EntityManager에서 contains 메서드를 사용

EntityManager의 contains 메서드를 사용하여 해당 엔티티가 현재 영속성 컨텍스트에 포함되어 있는지 확인할 수 있습니다. **새로운 엔티티는 영속성 컨텍스트에 포함되지 않으므로 false를 반환합니다.**

```java
@Autowired
private EntityManager entityManager;

public boolean isNewEntity(Object entity) {
    return !entityManager.contains(entity);
}
```


2. 엔티티의 식별자 필드를 확인  

만약 엔티티의 식별자를 개발자가 임의로 수동 할당하지 않는 경우라면,  
엔티티의 식별자(@Id) 가 null인지를 확인하여 판단할 수 도 있습니다. DB에 저장된 엔티티는 식별자값이 할당되므로 null일 수 없습니다.


3. Persistable 인터페이스 구현
JPA 2.0 버전에서는  Persistable 인터페이스를 구현하여 새로운 엔티티를 판단할 수 있습니다. 엔티티 클래스에 Persistable 인터페이스를 구현하고, isNew 메서드를 오버라이드하여 엔티티가 새로운지 여부를 정의할 수 있습니다.  
```java
@Entity
public class MyEntity implements Persistable<Long> {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    ....

    @Override
    public boolean isNew() {
        return id == null;
    }
}
```
4. @EntityListeners 사용   
JPA 엔티티 리스너를 사용하여 엔티티의 상태 변화를 감지할 수 있습니다. 이를 통해 엔티티가 새로운 엔티티인지 여부를 판단할 수 있습니다.  
```java
@Entity
@EntityListeners(MyEntityListener.class)
public class MyEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
   ....
}

public class MyEntityListener {

    @PrePersist
    public void prePersist(MyEntity entity) {
        // 새로운 엔티티가 영속화되기 전에 호출됨
        System.out.println("새로운 엔티티 입니다.");
    }
}
```

**Comment by GIVEN53** [comment]
새로운 Entity인지 여부는 JpaEntityInformation의 `isNew(T entity)`에 의해 판단됩니다.

다른 설정이 없으면 JpaEntityInformation의 구현체 중 JpaMetamodelEntityInformation 클래스가 동작합니다.

![image](https://github.com/2024-woowacourse-study/level-interview/assets/101033262/e8d007f0-8a29-456e-894e-b10b9fe3ed35)

`@Version`이 사용된 필드가 없거나 `@Version`이 사용된 필드가 primitive 타입이면 AbstractEntityInformation의 `isNew()`를 호출합니다.
`@Version`이 사용된 필드가 wrapper class이면 null여부를 확인합니다.

`@Version`이 사용된 필드가 없어서 AbstractEntityInformation 클래스가 동작하면

![image](https://github.com/2024-woowacourse-study/level-interview/assets/101033262/22e6ab92-f3d0-40a4-ac42-a30aca196d49)

`@Id` 어노테이션을 사용한 필드를 확인해서 primitive 타입이 아니라면 null 여부, Number의 하위 타입이면 0인지 여부를 확인합니다.

`@GeneratedValue` 어노테이션으로 키 생성 전략을 사용하면 데이터베이스에 저장될 때 id가 할당됩니다. 따라서 데이터베이스에 저장되기 전에 메모리에서 생성된 객체는 id가 비어있기 때문에 `isNew()`는 true가 되어 새로운 entity로 판단합니다.

키 생성 전략을 사용하지 않고 직접 ID를 할당하는 경우 새로운 entity로 간주되지 않습니다.
이 때는 엔티티에서 `Persistable<T>` 인터페이스를 구현해서 JpaMetamodelEntityInformation 클래스가 아닌 **JpaPersistableEntityInformation**의 `isNew()`가 동작하도록 합니다.

![image](https://github.com/2024-woowacourse-study/level-interview/assets/101033262/fd9e0dae-f777-442c-a725-5d8c07dd61d9)

**Comment by alstn113** [comment]
참고 자료: https://leegicheol.github.io/jpa/jpa-is-new/
